phoenix-core-client/src/main/java/org/apache/phoenix/schema/MetaDataClient.java

```diff
@@ -785,19 +785,37 @@ private boolean avoidRpcToGetTable(boolean alwaysHitServer, Long resolvedTimesta
             final long effectiveUpdateCacheFreq;
             final String ucfInfoForLogging; // Only used for logging purposes
 
+            boolean overrideUcfToDefault = false;
+            if (table.getType() == INDEX) {
+                overrideUcfToDefault =
+                        PIndexState.PENDING_DISABLE.equals(table.getIndexState()) ||
+                                !IndexMaintainer.sendIndexMaintainer(table);
+            }
+            if (!overrideUcfToDefault && !table.getIndexes().isEmpty()) {
+                List<PTable> indexes = table.getIndexes();
+                List<PTable> maintainedIndexes =
+                        Lists.newArrayList(IndexMaintainer.maintainedIndexes(indexes.iterator()));
+                // The maintainedIndexes contain only the indexes that are used by clients
+                // while generating the mutations. If all the indexes are usable by clients,
+                // we don't need to override UPDATE_CACHE_FREQUENCY. However, if any index is
+                // not in usable state by the client mutations, we should override
+                // UPDATE_CACHE_FREQUENCY to default value so that we make getTable() RPC calls
+                // until all index states change to ACTIVE, BUILDING or other usable states.
+                overrideUcfToDefault = indexes.size() != maintainedIndexes.size();
+            }
+
             // What if the table is created with UPDATE_CACHE_FREQUENCY explicitly set to ALWAYS?
             // i.e. explicitly set to 0. We should ideally be checking for something like
             // hasUpdateCacheFrequency().
 
             //always fetch an Index in PENDING_DISABLE state to retrieve server timestamp
             //QueryOptimizer needs that to decide whether the index can be used
-            if (PIndexState.PENDING_DISABLE.equals(table.getIndexState())) {
+            if (overrideUcfToDefault) {
```

Comment:
@virajjasani Default today is ALWAYS so this would work but what if default is something else? Should we explicitly override to ALWAYS? 
This is why I had added a separate config for `DEFAULT_UPDATE_CACHE_FREQUENCY_FOR_PENDING_DISABLED_INDEX` and set that to ALWAYS.

Reply:
It seems we are not directly using ALWAYS anywhere but get connection level prop: `phoenix.default.update.cache.frequency`. While if this is not set, obviously it gets default value of `ALWAYS`, but if `phoenix.default.update.cache.frequency` is set at Connection level, it means we need to use it. It basically means that the client really does not care for `ALWAYS` at all, it is risky for client to even use this prop at Connection level but whoever uses it, must know what consequences they will deal with.

Reply:
> This is why I had added a separate config for `DEFAULT_UPDATE_CACHE_FREQUENCY_FOR_PENDING_DISABLED_INDEX` and set that to ALWAYS.

Yes, i can remove this also with this PR.

Reply:
For PHOENIX-6883, we wanted to set default UCF to NEVER and that is why this config was introduced. At this point, I am a little confused about how this UCF change and PHOENIX-6883 would work together in the future.

Reply:
Exactly, we will need some level of re-iteration of PHOENIX-6883 with UCF. Ultimately, when client sets UCF to some value other than ALWAYS and NEVER, we need to ensure no getTable() or validateDdlTimestamp() RPC calls are initiated by clients for the duration of UCF. Its good that now we know this, so we will re-iterate on PHOENIX-6883 :)